File: src/Submission_In_Python/Lab1/Lab1.py
import sys
from queue import PriorityQueue

'''
    Please use Python 3.10.8 or higher
'''


class Node:
    def __init__(self):
        self.heuristic_value: int = int(1e9)
        self.adj_nodes: [tuple[str, int]] = []

    def __str__(self) -> str:
        return "heuristic_value " + str(self.heuristic_value) + " adj_nodes " + str(self.adj_nodes)


graph: dict[str: Node] = {}
distance: dict[str: int] = {}
parent: dict[str: str] = {}
input_lines: [str] = []


def taking_input_from_txt_file():
    with open("input.txt", "r") as file:
        for readline in file:
            input_lines.append(readline.strip())
            # Time to initialize the graph
            city_name: str = readline.strip().split()[0]
            graph[city_name] = Node()
            distance[city_name] = int(1e9)
            parent[city_name] = "null"
        file.close()  # Closing the input stream


taking_input_from_txt_file()


# Time to build the graph
def build_graph():
    for _i in input_lines:
        cur_line = _i.split()
        city_name = cur_line[0]
        heuristic_value = int(cur_line[1])
        graph.get(city_name).heuristic_value = heuristic_value
        for _j in range(2, len(cur_line), 2):
            graph.get(city_name).adj_nodes.append((cur_line[_j], int(cur_line[_j + 1])))


build_graph()


# Time to do a star search
def a_start_search(source: str, destination: str):
    distance[source] = 0
    final_cost = graph.get(source).heuristic_value
    pq: PriorityQueue[tuple[int, str, int]] = PriorityQueue()
    # putting in tuple in this format final_cost, city_name, distance/cost
    pq.put((final_cost, source, 0))
    while not pq.empty():
        cur_final_cost, cur_visiting_node, cur_distance = pq.get()
        if cur_visiting_node == destination:
            break
        if distance.get(cur_visiting_node) != cur_distance:
            continue
        for k in graph.get(cur_visiting_node).adj_nodes:
            if distance.get(k[0]) > cur_distance + k[1]:
                distance[k[0]] = cur_distance + k[1]
                parent[k[0]] = cur_visiting_node
                update_final_cost = distance.get(k[0]) + graph.get(k[0]).heuristic_value
                pq.put((update_final_cost, k[0], distance.get(k[0])))


starting_node = input("Start node: ")
ending_node = input("Destination: ")
a_start_search(starting_node, ending_node)
if distance.get(ending_node) == int(1e9):
    print("NO PATH FOUND")
    sys.exit(0)
print(distance.get(ending_node))
path: [str] = []
cur_node = ending_node
while cur_node != starting_node:
    path.append(cur_node)
    cur_node = parent.get(cur_node)
path.append(starting_node)
path.reverse()
for i in range(len(path) - 1):
    print(path[i] + " -> ", end="")
print(path[len(path) - 1])

File: src/Submission_In_Java/Lab1/Lab1.java
package Submission_In_Java.Lab1;

import java.io.*;
import java.util.*;

public class Lab1 {
    private static final PrintWriter pw = new PrintWriter(new BufferedOutputStream(System.out));
    private static BufferedReader br;

    static {
        try {
            br = new BufferedReader(new InputStreamReader(new FileInputStream("~ path to your input.txt")));
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    private static final HashMap<String, Node> graph = new HashMap<>();
    private static final ArrayList<String> inputLines = new ArrayList<>();
    private static final HashMap<String, Integer> distance = new HashMap<>();
    private static final HashMap<String, String> parent = new HashMap<>();

    public static void main(String[] args) throws IOException {
        // Take input file
        takeInputFromTxtFile();
        // Initialize graph
        initializeGraph();
        // Build the graph
        buildGraph();
        // Need to take input from the console
        // So re-assigning it to the std i/o
        br = new BufferedReader(new InputStreamReader(System.in));
        pw.print("Start Node: ");
        pw.flush();
        String source = new StringTokenizer(br.readLine()).nextToken();
        pw.print("Destination: ");
        pw.flush();
        String destination = new StringTokenizer(br.readLine()).nextToken();
        br.close(); // Closing the input stream
        // Need to write A*
        aStarSearch(source, destination);
        if (distance.get(destination) == Integer.MAX_VALUE) {
            pw.println("No path found");
            System.exit(0);
        }
        pw.println(distance.get(destination));
        ArrayList<String> path = new ArrayList<>();
        String curCity = destination;
        path.add(curCity);
        while (!curCity.equals(source)) {
            curCity = parent.get(curCity);
            path.add(curCity);
        }
        //noinspection DuplicatedCode
        Collections.reverse(path);
        for (int i = 0; i < path.size() - 1; i++) {
            pw.print(path.get(i) + " -> ");
        }
        pw.println(path.get(path.size() - 1));
        pw.close();
    }

    private static void aStarSearch(String source, String destination) {
        distance.put(source, 0);
        var sourceFinalCost = graph.get(source).heuristicValue;
        PriorityQueue<StringIntIntPair> pq = new PriorityQueue<>(Comparator.comparingInt(o -> o.third));
        pq.add(new StringIntIntPair(source, 0, sourceFinalCost));
        while (!pq.isEmpty()) {
            var curCity = pq.peek().first;
            var curDistance = pq.peek().second;
            pq.poll();
            if (curCity.equals(destination)) break;
            if (distance.get(curCity) != curDistance) continue;
            for (var k : graph.get(curCity).adjNodes) {
                if (distance.get(k.first) > curDistance + k.second) {
                    distance.put(k.first, curDistance + k.second);
                    parent.put(k.first, curCity);
                    var curFinalCost = distance.get(k.first) + graph.get(k.first).heuristicValue;
                    pq.add(new StringIntIntPair(k.first, distance.get(k.first), curFinalCost));
                }
            }
        }
    }


    private static void buildGraph() {
        for (var l : inputLines) {
            StringTokenizer tk = new StringTokenizer(l);
            String cityName = tk.nextToken();
            var curNode = graph.get(cityName);
            curNode.heuristicValue = Integer.parseInt(tk.nextToken());
            while (tk.countTokens() > 0) {
                String adjCity = tk.nextToken();
                int distance = Integer.parseInt(tk.nextToken());
                curNode.adjNodes.add(new StringIntPair(adjCity, distance));
                graph.get(adjCity).adjNodes.add(new StringIntPair(cityName, distance));
            }
        }
    }

    private static void initializeGraph() {
        for (var l : inputLines) {
            StringTokenizer tk = new StringTokenizer(l);
            var cityName = tk.nextToken();
            graph.put(cityName, new Node());
            distance.put(cityName, (int) 1e9);
            parent.put(cityName, null);
        }
    }

    private static void takeInputFromTxtFile() throws IOException {
        String inputLine;
        while ((inputLine = br.readLine()) != null) {
            inputLines.add(inputLine);
        }
    }

    private record StringIntPair(String first, int second) {
    }

    private record StringIntIntPair(String first, int second, int third) {
    }

    private static class Node {
        private int heuristicValue;
        private final List<StringIntPair> adjNodes = new ArrayList<>();
    }
}
